### DISCUSSION TITLE
[allure-cucumber7-jvm] Is it possible to add attachment(screenshot) to each step?

### DISCUSSION THREAD

**Discussion by ivanbraynov** [allure-framework/allure2#1741] [Questions & Support] [answered]
Hello there,

I am using Java/Selenium/CucumberJvm7/TestNG/Maven for an automations framework.
The issue I have is that when I add a screenshot in `@AfterStep` it goes in the tear down section. This is kinda troublesome when you have a lot of steps and you want to debug the ui state prior the failed step. I have to expand every tear down child and click on the screenshot to see if that is the one I am looking for. If I had it attached to every step it would be perfect.

For attaching the screenshot I use:
```
    @AfterStep
    public void addAllureScreenshot(Scenario scenario) throws FileNotFoundException {
        logger.info("after step add screenshot");
        Allure.addAttachment("screenshot", new ByteArrayInputStream(((TakesScreenshot) driver).getScreenshotAs(OutputType.BYTES)));
    }
```

![image](https://user-images.githubusercontent.com/49759194/186184480-1dc59057-f8ef-43ac-9a1d-7d83ada23ed9.png)

I don't know if this is specific only to the cucumberjvm, because I've seen it working with testng only
![image](https://user-images.githubusercontent.com/49759194/186187155-e02fd95b-8cbe-4212-bef0-74be4a8f6a98.png)

Thank you!

**Comment by natfatykhova** [comment]
Hi Ivan, did you manage to solve the issue?

**Reply by ivanbraynov** [reply]
Hey, I did not find a solution.

**Comment by baev** [comment, marked as answer]
you can use `io.qameta.allure.listener.StepLifecycleListener#beforeStepStop` for that. 

```java
import io.qameta.allure.Allure;
import io.qameta.allure.listener.StepLifecycleListener;
import io.qameta.allure.model.StepResult;

public class StepScreenshotListener implements StepLifecycleListener {

    @Override
    public void beforeStepStop(final StepResult result) {
        // you can check the status of the step by accessing result.getStatus() if you need it


        Allure.addAttachment(
                "screenshot",
                new ByteArrayInputStream(((TakesScreenshot) driver).getScreenshotAs(OutputType.BYTES))
        );
    }
}
```

`StepScreenshotListener` is configured via Java SPI (you need to add `META-INF/services/io.qameta.allure.listener.StepLifecycleListener` file to your test resources folder and specify the full name of  `StepScreenshotListener` in it)
